File: Kotlin/RadixSort.kt
// Radix sort Java implementation
import java.io.*
import java.util.*

internal class RadixSort {

    // A utility function to get maximum value in arr[]
    fun getMax(arr: IntArray, n: Int): Int {
        var mx = arr[0]
        for (i in 1 until n)
            if (arr[i] > mx)
                mx = arr[i]
        return mx
    }

    // A function to do counting sort of arr[] according to
    // the digit represented by exp.
    fun countSort(arr: IntArray, n: Int, exp: Int) {
        val output = IntArray(n) // output array
        var i: Int
        val count = IntArray(10)
        Arrays.fill(count, 0)

        // Store count of occurrences in count[]
        i = 0
        while (i < n) {
            count[arr[i] / exp % 10]++
            i++
        }

        // Change count[i] so that count[i] now contains
        // actual position of this digit in output[]
        i = 1
        while (i < 10) {
            count[i] += count[i - 1]
            i++
        }

        // Build the output array
        i = n - 1
        while (i >= 0) {
            output[count[arr[i] / exp % 10] - 1] = arr[i]
            count[arr[i] / exp % 10]--
            i--
        }

        // Copy the output array to arr[], so that arr[] now
        // contains sorted numbers according to curent digit
        i = 0
        while (i < n) {
            arr[i] = output[i]
            i++
        }
    }

    // The main function to that sorts arr[] of size n using
    // Radix Sort
    fun radixsort(arr: IntArray, n: Int) {
        // Find the maximum number to know number of digits
        val m = getMax(arr, n)

        // Do counting sort for every digit. Note that instead
        // of passing digit number, exp is passed. exp is 10^i
        // where i is current digit number
        var exp = 1
        while (m / exp > 0) {
            countSort(arr, n, exp)
            exp *= 10
        }
    }

    // A utility function to print an array
    fun print(arr: IntArray, n: Int) {
        for (i in 0 until n)
            System.out.print(arr[i].toString() + " ")
    }


    /*Driver function to check for above function*/
    fun main(args: Array<String>) {
        val arr = intArrayOf(23, 44, 2, 1, 2334, 2)
        val n = arr.size
        radixsort(arr, n)
        print(arr, n)
    }
}

File: Cpp/RadixSort.cpp
#include <iostream>
using namespace std;

//maximum value in arr[ ]
int getMax(int arr[], int n)
{
	int mx = arr[0];
	for (int i = 1; i < n; i++)
		if (arr[i] > mx)
			mx = arr[i];
	return mx;
}

//counting sort of arr[ ] according to exp
void countSort(int arr[], int n, int exp)
{
    //output
	int op[n];
	int i, count[10] = { 0 };

	for (i = 0; i < n; i++)
		count[(arr[i] / exp) % 10]++;

	//position of this digit in op[ ]
	for (i = 1; i < 10; i++)
		count[i] += count[i - 1];

	for (i = n - 1; i >= 0; i--) {
		op[count[(arr[i] / exp) % 10] - 1] = arr[i];
		count[(arr[i] / exp) % 10]--;
	}

	//make arr[ ] contain sorted numbers according to current digit
	for (i = 0; i < n; i++)
		arr[i] = op[i];
}

void radixsort(int arr[], int n)
{
	int m = getMax(arr, n);

    //exp is 10^i where i is current digit number
	for (int exp = 1; m / exp > 0; exp *= 10)
		countSort(arr, n, exp);
}

void print(int arr[], int n)
{
	for (int i = 0; i < n; i++)
		cout << arr[i] << " ";
}

int main()
{
	int arr[] = { 170, 45, 75, 90, 802, 24, 2, 66 };
	int n = sizeof(arr) / sizeof(arr[0]);
	
	radixsort(arr, n);
    cout<<" The sorted array is: \n";
	print(arr, n);
	return 0;
}